include/swift/AST/ConstTypeInfo.h

```diff
@@ -199,6 +200,37 @@ class TypeValue : public CompileTimeValue {
   swift::Type Type;
 };
 
+/// A representation of a Keypath
+class KeyPathValue : public CompileTimeValue {
+public:
+  struct Component {
+    std::string Label;
+    swift::Type Type;
+  };
+  KeyPathValue(std::string Path,
+               swift::Type RootType,
+               swift::Type ValueType,
+               std::vector<Component> Components)
+  : CompileTimeValue(ValueKind::KeyPath), Path(Path), RootType(RootType), ValueType(ValueType), Components(Components) {}
+
+  std::string getPath() const { return Path; }
+  swift::Type getRootType() const { return RootType; }
+  swift::Type getValueType() const { return ValueType; }
+  std::vector<Component> getComponents() const {
+    return Components;
+  }
+
+  static bool classof(const CompileTimeValue *T) {
+    return T->getKind() == ValueKind::KeyPath;
+  }
+
+private:
+  std::string Path;
```

Comment:
Similarly, can this instead be computed by iterating over the components? That would make `extractNestedName` unnecessary.

Reply:
Yeah, that is definitely possible. I just wasn't sure if `->getComponents()` has the following characteristics:
1. It is guaranteed to follow the path order
2. There are no cases where `component.isResolved()` would be false, causing certain path components to not be populated. 

@artemcm Do you know if these are assumptions we can make?

Reply:
I am sure we should be able to make (1) assumption. But if (2) is false for the last component, I think both mechanisms this code uses would still yield the same result? So it should be safe to make as well.

Reply:
Removed extractNestedNames and updated the path to be computed from the components